File: Week_09/32/longest_valid_parentheses2.go
package main

func longestValidParentheses2(s string) int {
	maxLen := 0
	stack := []int{-1}
	for i := 0; i < len(s);i++ {
		if s[i] == '(' {
			stack = append(stack, i)
		} else {
			stack = stack[:len(stack)-1]
			if len(stack) == 0 {
				stack = append(stack, i)
			} else {
				maxLen = max(maxLen, i-stack[len(stack)-1])
			}
		}
	}
	return maxLen
}

File: Week_09/32/longest_valid_parentheses2.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        ans = 0
        stack = [-1]
        for i in range(len(s)):
            if s[i] == '(':
                stack.append(i)
            else:
                stack.pop()
                if len(stack) == 0:
                    stack.append(i)
                else:
                    ans = max(ans, i-stack[-1])
        return ans

    def test(self):
        s = "(())"
        ans = self.longestValidParentheses(s)
        target = 4
        assert ans == target, f"target: {target}, ans: {ans}"


if __name__ == "__main__":
    Solution().test()
